README.md
# Statusmap panel for Grafana

[![Download from GitHub](https://img.shields.io/github/tag-date/flant/grafana-statusmap.svg?logo=github&label=latest)](https://github.com/flant/grafana-statusmap/releases/latest)
[![GH Discussions](https://img.shields.io/badge/GitHub-discussions-brightgreen)](https://github.com/flant/grafana-statusmap/discussions)
[![Telegram chat RU](https://img.shields.io/badge/@statusmap_ru-RU-informational.svg?logo=telegram)](https://t.me/statusmap_ru)
[![Follow Flant on Twitter](https://img.shields.io/twitter/follow/flant_com?label=Follow%20Flant&style=social)](https://twitter.com/flant_com)

Panel to show discrete statuses of multiple targets over time.

![Statusmap sample panel](https://raw.githubusercontent.com/flant/grafana-statusmap/master/docs/img/flant-statusmap-panel.png)

![Statusmap sample panel with dark theme](https://raw.githubusercontent.com/flant/grafana-statusmap/master/docs/img/flant-statusmap-panel-dark.png)

> Run `docker compose up` and open http://localhost:3000 in browser to see a simple demo.


## Features

* Grouping values into rows and buckets using legend from query
* User defined color mapping
* Multiple values in bucket are displayed via tooltip
* Configurable tooltip items
* Pagination for rows
* Increasing rows/buckets' interval for better visual representation
* Representing null values as empty bucket or zero value

### Supported environment

* Tested with:
  - Prometheus
  - InfluxDB
  - Mysql
  - Grafana 7.x, 8.x
* Support for earlier Grafana versions:
  - 0.5.x versions should work in Grafana 6.5.x - 6.7.x with timeseries data format
  - (0.4.2)[https://github.com/flant/grafana-statusmap/releases/tag/v0.4.2] should work in Grafana 5.4.x - 6.4.x

## Motivation

We had a desperate need to visualize a set of timeseries statuses over time period, so we can
see a history of changes for objects' status. Since we maintain a lot of Kubernetes clusters
(and related infrastructure), our main cases for that are visualization of servers & Kubernetes
pods health states as well as HTTP services health checks. We've tried a variety of Grafana
plugins available (they are listed in *Acknowledgements* below) but none of them could provide
the features and visualization really close to what we've been looking for.

_NB: You can find more details about our journey of creating the plugin in
[this post](https://medium.com/flant-com/statusmap-grafana-plugin-to-visualize-status-over-time-fe6ced391853)._

_Objects_ being visualized with this plugin may be different: not only IT components (e.g. server
hosts and Kubernetes pods) but just anything you can imagine like coffee makers on the picture
above. These objects should have _discrete statuses_ which are sets of predefined values, e.g.
`ok` = 0, `off` = 1, `fail` = 2, etc.

## Configuration

### Datasource notes

To create neat graphs your datasource should return good data. Plugin adjust `$__interval` variable depending on
bucket width in panel options. Your queries should aggregate statuses over `$__interval`.

To make multiple values mode works as expected you should define multiple queries: one query for each possible status.

Plugin doesn't aggregate data in time for now, it only renders input data as buckets. Because of this
data should contain points for each timestamp in time range and equal timestamps for every possible
target (y-axis label). This limitation is addressed by [issue #53](https://github.com/flant/grafana-statusmap/issues/53).

#### Prometheus

To work with data from Prometheus you will need to setup discrete statuses for your objects.
Requirements to store these statuses in metrics are as follows:
* metrics should have two values: `0` and `1`;
* there should be a label with status' value.

When it's done, you can collect all the data via query, e.g.:

```
(max_over_time(coffee_maker_status{status="<STATUS_VALUE>"}[$__interval]) == 1) * <STATUS_VALUE>
```

If there was no such status (`<STATUS_VALUE>`) during query's interval, Prometheus would
return nothing. Otherwise, status' value will be returned.

For example, if you have 5 types of statuses and a metric (`coffee_maker_status`) with 5
allowed values (`0`, `1`, `2`, `3`, `4`), you should transform this metric using following rule:

```
- record: coffee_maker_status:discrete
  expr: |
    count_values("status", coffee_maker_status)
```

That's how `coffee_maker_status` metric with value `3` will be transformed into new metric:

```
coffee_maker_status:discrete{status="3"} 1
```

Now, when Prometheus has `0` and `1` values for each status, all these metrics can be
aggregated, so you will get all available statuses of your objects over time.

#### InfluxDB

Choose 'Time series' for 'Format as' and use `GROUP BY ($__interval)` in query. `$tag_<tag name>` can be used in 'Alias by' to define y-axis labels.

#### Mysql

Example query with aggregation over `$__interval` is like this (you need one query for each possible status value):

```
SELECT
  $__timeGroupAlias(date_insert,$__interval),
  name AS metric,
  min(statusi) AS "statusi"
FROM coffee_makers
WHERE
  $__timeFilter(date_insert) AND statusi=1
GROUP BY 1,2
ORDER BY $__timeGroup(date_insert,$__interval)
```

`metric` column is used as y-axis label.

### Panel

First of all, an individual query for each possible status' value should be created.
Each query should also have similar legend for grouping:

![Query setup](https://raw.githubusercontent.com/flant/grafana-statusmap/master/docs/img/queries-example.png)

Then, color mapping for status' values should be defined in __Discrete__ color mode:

![Color mapping](https://raw.githubusercontent.com/flant/grafana-statusmap/master/docs/img/color-mapping.png)

Use can use presets to define a trafic light colors or 8 colors from [solarized](https://ethanschoonover.com/solarized/) palette:

![Color mapping empty](https://raw.githubusercontent.com/flant/grafana-statusmap/master/docs/img/color-preset-01.png)

![Color mapping trafic lights](https://raw.githubusercontent.com/flant/grafana-statusmap/master/docs/img/color-preset-02.png)


_Note: __Spectrum__ and __Opacity__ color modes function the same way they do in [Heatmap](https://grafana.com/plugins/heatmap) plugin._


### More options

#### Bucket

![Bucket options](https://raw.githubusercontent.com/flant/grafana-statusmap/master/docs/img/options-bucket.png)

__Multiple values__ checkbox specifies how they should be displayed:
* If it's off, multiple values for one bucket are treated as error;
* If it's on, color for such bucket would be determined by the value having least index in color mapping.

![Color mapping](https://raw.githubusercontent.com/flant/grafana-statusmap/master/docs/img/multiple-values-error.png)

__Display nulls__ can be treated as empty buckets or displayed with the color of `0` value.

![Color mapping](https://raw.githubusercontent.com/flant/grafana-statusmap/master/docs/img/null-as-empty.png)

__Min width__ and __spacing__ are used to specify minimal bucket width and spacing between buckets.
__Rounding__ may be used to round edges.

![Min width, spacing, rounding 1](https://raw.githubusercontent.com/flant/grafana-statusmap/master/docs/img/min-width-spacing-rounding-01.png)

![Min width, spacing, rounding 2](https://raw.githubusercontent.com/flant/grafana-statusmap/master/docs/img/min-width-spacing-rounding-02.png)

__Values index__ set to positive number to display only values from specified timeseries.

#### Display

![Display options](https://raw.githubusercontent.com/flant/grafana-statusmap/master/docs/img/options-display.png)

__Show legend__ checkbox toggles legend at the bottom of the panel.

__Rows sort__ can be used to sort labels on Y axis. Metrics — sort y labels as they are defined on Metrics tab. a→z and z→a sort labels descending or ascending in a [natural](https://developer.mozilla.org/ru/docs/Web/JavaScript/Reference/Global_Objects/String/localeCompare) order.

#### Pagination

![Pagination controls](https://raw.githubusercontent.com/flant/grafana-statusmap/master/docs/img/pagination-graph.png)

__Enable pagination__ toggles pagination controls on graph.

__Rows per page__ a number of rows to display on graph.

#### Tooltip

![Tooltip in frozen state](https://raw.githubusercontent.com/flant/grafana-statusmap/master/docs/img/tooltip-frozen.png)

__Show tooltip__ toggles tooltip display on mouse over buckets.

__Freeze on click__ toggles tooltip "freezing" on click. Frozen tooltip can be used to compare data with floating tooltip or to follow URLs.


CHANGELOG.md
# Changelog

## v0.5.1

- Fix fallback to timeseries for Grafana 6.x
- Change docker-compose.yaml to work with Grafana 6.x

## v0.5.0

- Support DataFrames. (#239)
- Add tooltip variables with percentual bucket range: __bucket_from, __bucket_to (#242)
- Support Grafana 8.4+
- Hide markers in tooltip.

## v0.4.2

- Compatibility with Grafana 8.1+. (#230, #219, #220, #221)
- Fix 'as zero' for spectrum mode. (#208, #209)

## v0.4.1

- Fix regression in discrete mode. "Data has value with undefined color"

## v0.4.0

- Add __y_label_trim variable for tooltip items.
- Add y-label splitter for tooltip items.
- Add custom content for tooltip.
- Fix to work with Grafana 7.4.x.
- Migrate to grafana-toolkit and yarn to build and sign.
- Cleanup dist directory content, reduce compiled plugin size, remove dist from repository.
- Code cleanup: fix huge amount of linter errors and warnings.
- Add Github Actions configuration to automate release process. 
- Add unobtrusive branding.

## v0.3.4

- Fix to work in Grafana 7.2.0,7.2.1
- Fix 'name' in tooltip for opacity and gradient modes
- Fix color scale for opacity and gradient modes

## v0.3.3

- Fix fallback to string events for older versions of Grafana

## v0.3.2

- Fix plugin metadata for Grafana repository

## v0.3.1

- Fix to work in Grafana 7.1.0

## v0.3.0

- Tooltip freeze on click
- Show URLs in tooltip
- Pagination controls
- Fix tooltip and null buckets
- Fix display for 6.7+, 7.0+

## v0.2.0

- Migrate code base to typescript
- Fixes to work in 6.3.0+ grafana

## v0.1.1

- Fix for annotations in grafana 5.x before 5.4

## v0.1.0

- Tested with Grafana 6.0.0
- Tested with InfluxDB and Mysql datasources
- Add initial support for display annotations
- Add example for k8s statuses (thanks, @vrutkovs)
- Fix hanging on big values
- Fix horizontal spacing = 0
- Fix for "Object doesn't support property or method 'remove'"
- Fix card width for targets with different datapoints count

## v0.0.4

- Fix display of multivalues buckets as an empty cell

## v0.0.3

- Add solarized preset
- Reorganize statuses editor for discrete mode
- Separate options for vertical and horizontal spacing for cards
- Add simple sort options for Y axis
- Fix display null values as zero

## v0.0.2

- Install with GF_INSTALL_PLUGINS
- Fix legend overlap
- Fix colors for dark theme
- Fix panel rendering timeout error


## v0.0.1

- First public release



src/module.ts
// Libraries
import _ from 'lodash';
import { auto } from 'angular';

import { PanelEvents, DataFrame } from '@grafana/data';
import TimeSeries from 'grafana/app/core/time_series2';

import { config } from '@grafana/runtime';
import { DataProcessor } from './data_processor';

// Components
import './color_legend';
import { optionsEditorCtrl } from './options_editor';
import { tooltipEditorCtrl } from './tooltip_editor';
import { migratePanelConfig } from './panel_config_migration';

// Utils
import kbn from 'grafana/app/core/utils/kbn';
import { loadPluginCss, MetricsPanelCtrl } from 'grafana/app/plugins/sdk';

// Types
import { AnnotationsSrv } from 'grafana/app/features/annotations/annotations_srv';
import { Bucket, BucketMatrix, BucketMatrixPager } from './statusmap_data';
import rendering from './rendering';

import { ColorModeDiscrete } from './color_mode_discrete';
import { CoreEvents } from './util/grafana/events';

const VALUE_INDEX = 0,
  TIME_INDEX = 1;

const colorSchemes = [
  // Diverging
  { name: 'Spectral', value: 'interpolateSpectral', invert: 'always' },
  { name: 'RdYlGn', value: 'interpolateRdYlGn', invert: 'always' },
  { name: 'GnYlRd', value: 'interpolateGnYlRd', invert: 'always' },

  // Sequential (Single Hue)
  { name: 'Blues', value: 'interpolateBlues', invert: 'dark' },
  { name: 'Greens', value: 'interpolateGreens', invert: 'dark' },
  { name: 'Greys', value: 'interpolateGreys', invert: 'dark' },
  { name: 'Oranges', value: 'interpolateOranges', invert: 'dark' },
  { name: 'Purples', value: 'interpolatePurples', invert: 'dark' },
  { name: 'Reds', value: 'interpolateReds', invert: 'dark' },

  // Sequential (Multi-Hue)
  { name: 'BuGn', value: 'interpolateBuGn', invert: 'dark' },
  { name: 'BuPu', value: 'interpolateBuPu', invert: 'dark' },
  { name: 'GnBu', value: 'interpolateGnBu', invert: 'dark' },
  { name: 'OrRd', value: 'interpolateOrRd', invert: 'dark' },
  { name: 'PuBuGn', value: 'interpolatePuBuGn', invert: 'dark' },
  { name: 'PuBu', value: 'interpolatePuBu', invert: 'dark' },
  { name: 'PuRd', value: 'interpolatePuRd', invert: 'dark' },
  { name: 'RdPu', value: 'interpolateRdPu', invert: 'dark' },
  { name: 'YlGnBu', value: 'interpolateYlGnBu', invert: 'dark' },
  { name: 'YlGn', value: 'interpolateYlGn', invert: 'dark' },
  { name: 'YlOrBr', value: 'interpolateYlOrBr', invert: 'dark' },
  { name: 'YlOrRd', value: 'interpolateYlOrRd', invert: 'dark' },
];

let colorModes = ['opacity', 'spectrum', 'discrete'];
let opacityScales = ['linear', 'sqrt'];

loadPluginCss({
  dark: 'plugins/flant-statusmap-panel/styles/dark.css',
  light: 'plugins/flant-statusmap-panel/styles/light.css',
});

class StatusHeatmapCtrl extends MetricsPanelCtrl {
  static templateUrl = 'module.html';

  data: any;
  bucketMatrix: BucketMatrix;
  bucketMatrixPager: BucketMatrixPager;

  graph: any;
  opacityScales: any = [];
  colorModes: any = [];
  colorSchemes: any = [];
  unitFormats: any;

  dataWarnings: { [warningId: string]: { title: string; tip: string } } = {};
  multipleValues: boolean;
  noColorDefined: boolean;
  noDatapoints: boolean;

  discreteExtraSeries: ColorModeDiscrete;

  annotations: object[] = [];
  annotationsPromise: any;

  // TODO remove this transient variable: use ng-model-options="{ getterSetter: true }"
  pageSizeViewer = 15;

  panelDefaults: any = {
    // datasource name, null = default datasource
    datasource: null,
    // color mode
    color: {
      mode: 'spectrum',
      cardColor: '#b4ff00',
      colorScale: 'sqrt',
      exponent: 0.5,
      colorScheme: 'interpolateGnYlRd',
      // discrete mode settings
      defaultColor: '#757575',
      thresholds: [], // manual colors
    },
    // buckets settings
    cards: {
      cardMinWidth: 5,
      cardVSpacing: 2,
      cardHSpacing: 2,
      cardRound: null,
    },
    xAxis: {
      show: true,
    },
    yAxis: {
      show: true,
      minWidth: -1,
      maxWidth: -1,
    },
    tooltip: {
      show: true,
      freezeOnClick: true,
      showItems: false,
      items: [], // see tooltip_editor.ts
      showExtraInfo: false,
      extraInfo: '',
    },
    legend: {
      show: true,
    },
    yLabel: {
      usingSplitLabel: false,
      delimiter: '',
      labelTemplate: '',
    },
    // how null points should be handled
    nullPointMode: 'as empty',
    yAxisSort: 'metrics',
    highlightCards: true,
    useMax: true,

    seriesFilterIndex: -1,

    // Pagination options
    usingPagination: false,
    pageSize: 15,

    // Branding flag.
    hideBranding: false,
  };

  processor?: DataProcessor;

  /** @ngInject */
  constructor($scope: any, $injector: auto.IInjectorService, private annotationsSrv: AnnotationsSrv) {
    super($scope, $injector);

    // Grafana 7.2 workaround
    if (typeof kbn['intervalToMs'] === 'function') {
      kbn.interval_to_ms = kbn.intervalToMs;
    }

    migratePanelConfig(this.panel);
    _.defaultsDeep(this.panel, this.panelDefaults);

    this.bucketMatrix = new BucketMatrix();

    // Create pager for bucketMatrix and restore page size.
    this.bucketMatrixPager = new BucketMatrixPager();
    this.bucketMatrixPager.setEnable(this.panel.usingPagination);
    this.bucketMatrixPager.setDefaultPageSize(this.panel.pageSize);
    this.bucketMatrixPager.setPageSize(this.panel.pageSize);
    $scope.pager = this.bucketMatrixPager;

    this.opacityScales = opacityScales;
    this.colorModes = colorModes;
    this.colorSchemes = colorSchemes;

    // default graph width for discrete card width calculation
    this.graph = {
      chartWidth: -1,
    };

    this.multipleValues = false;
    this.noColorDefined = false;

    this.discreteExtraSeries = new ColorModeDiscrete($scope);

    this.dataWarnings = {
      noColorDefined: {
        title: 'Data has value with undefined color',
        tip: 'Check metric values, color values or define a new color',
      },
      multipleValues: {
        title: 'Data has multiple values for one target',
        tip: 'Change targets definitions or set "use max value"',
      },
      noDatapoints: {
        title: 'No data points',
        tip: 'No datapoints returned from data query',
      },
    };

    this.annotations = [];
    this.annotationsSrv = annotationsSrv;

    this.events.on(PanelEvents.render, this.onRender.bind(this));
    this.events.on(PanelEvents.dataError, this.onDataError.bind(this));
    this.events.on(PanelEvents.editModeInitialized, this.onInitEditMode.bind(this));
    this.events.on(PanelEvents.refresh, this.postRefresh.bind(this));
    // custom event from rendering.js
    this.events.on(CoreEvents.renderComplete, this.onRenderComplete.bind(this));

    this.onCardColorChange = this.onCardColorChange.bind(this);

    console.log('Grafana buildInfo:', config.buildInfo);

    const majorVersion = parseInt(config.buildInfo.version.split('.', 2)[0], 10);
    if (majorVersion >= 7) {
      // Support data frames for 7.0+
      (this as any).useDataFrames = true;
      this.events.on(PanelEvents.dataFramesReceived, this.onDataFramesReceived.bind(this));
      this.events.on(PanelEvents.dataSnapshotLoad, this.onDataFramesReceived.bind(this));
      this.processor = new DataProcessor(this.panel, majorVersion);
    } else {
      // Fallback to support only timeseries data format.
      this.events.on(PanelEvents.dataReceived, this.onDataReceived.bind(this));
      this.events.on(PanelEvents.dataSnapshotLoad, this.onDataReceived.bind(this));
    }
  }

  onRenderComplete(data: any): void {
    this.graph.chartWidth = data.chartWidth;
    this.renderingCompleted();
  }

  changeDefaultPaginationSize(defaultPageSize: number): void {
    this.bucketMatrixPager.setDefaultPageSize(defaultPageSize);
    this.bucketMatrixPager.setPageSize(defaultPageSize);
    this.pageSizeViewer = defaultPageSize;

    this.render();
    this.refresh();
  }

  onChangePageSize(): void {
    if (this.pageSizeViewer <= 0) {
      this.pageSizeViewer = this.bucketMatrixPager.defaultPageSize;
    }
    this.bucketMatrixPager.setPageSize(this.pageSizeViewer);
    this.bucketMatrixPager.setCurrent(0);

    this.render();
    this.refresh();
  }

  onPrevPage(): void {
    this.bucketMatrixPager.switchToPrev();
    this.render();
  }

  onNextPage(): void {
    this.bucketMatrixPager.switchToNext();
    this.render();
  }

  // getChartWidth returns an approximation of chart canvas width or
  // a saved value calculated during a render.
  getChartWidth(): number {
    if (this.graph.chartWidth > 0) {
      return this.graph.chartWidth;
    }

    const wndWidth = $(window).width();
    // gripPos.w is a width in grid's measurements. Grid size in Grafana is 24.
    const panelWidthFactor = this.panel.gridPos.w / 24;
    const panelWidth = Math.ceil(wndWidth * panelWidthFactor);
    // approximate width of the chart draw canvas:
    // - y axis ticks are not rendered yet on first data receive,
    //   so choose 200 as a decent value for y legend width
    // - chartWidth can not be lower than the half of the panel width.
    const chartWidth = _.max([panelWidth - 200, panelWidth / 2]);

    return chartWidth!;
  }

  // Quick workaround for 6.7 and 7.0+. There is no call to
  // calculateInterval in updateTimeRange in those versions.
  // TODO ts type has no argument for this method.
  //
  updateTimeRange(datasource?: any) {
    // @ts-ignore
    let ret = super.updateTimeRange(datasource);
    this.calculateInterval();
    return ret;
  }

  // calculateInterval is called on 'refresh' to calculate an interval
  // for datasource.
  // It is override of calculateInterval from MetricsPanelCtrl.
  calculateInterval() {
    let chartWidth = this.getChartWidth();

    let minCardWidth = this.panel.cards.cardMinWidth;
    let minSpacing = this.panel.cards.cardHSpacing;
    let maxCardsCount = Math.ceil((chartWidth - minCardWidth) / (minCardWidth + minSpacing));

    let intervalMs;
    let rangeMs = this.range.to.valueOf() - this.range.from.valueOf();

    // this is minimal interval! kbn.round_interval will lower it.
    intervalMs = this.discreteExtraSeries.roundIntervalCeil(rangeMs / maxCardsCount);

    // Calculate low limit of interval
    let lowLimitMs = 1; // 1 millisecond default low limit

    let intervalOverride = this.panel.interval;

    // if no panel interval check datasource
    if (intervalOverride) {
      intervalOverride = this.templateSrv.replace(intervalOverride, this.panel.scopedVars);
    } else if (this.datasource && this.datasource.interval) {
      intervalOverride = this.datasource.interval;
    }

    if (intervalOverride) {
      if (intervalOverride[0] === '>') {
        intervalOverride = intervalOverride.slice(1);
      }
      lowLimitMs = kbn.interval_to_ms(intervalOverride);
    }

    if (lowLimitMs > intervalMs) {
      intervalMs = lowLimitMs;
    }
    let interval = kbn.secondsToHms(intervalMs / 1000);

    this.intervalMs = intervalMs;
    this.interval = interval;

    // Get final buckets count after interval is adjusted
    // TODO is it needed?
    //this.xBucketsCount = Math.floor(rangeMs / intervalMs);
  }

  issueQueries(datasource: any) {
    // Grafana 8.1+: there is no updateTimeRange call before initial issueQueries.
    // https://github.com/grafana/grafana/commit/6f38883583c4c43af149f68db482b39a3240ec95
    if (!this.range) {
      this.updateTimeRange(datasource);
    }
    this.annotationsPromise = this.annotationsSrv.getAnnotations({
      dashboard: this.dashboard,
      panel: this.panel,
      range: this.range,
    });

    /* Wait for annotationSrv requests to get datasources to
     * resolve before issuing queries. This allows the annotations
     * service to fire annotations queries before graph queries
     * (but not wait for completion). This resolves
     * issue 11806.
     */
    // 5.x before 5.4 doesn't have datasourcePromises.
    // @ts-ignore
    if ('undefined' !== typeof this.annotationsSrv.datasourcePromises) {
      // @ts-ignore
      return this.annotationsSrv.datasourcePromises.then(r => {
        return this.issueQueriesWithInterval(datasource, this.interval);
      });
    } else {
      return this.issueQueriesWithInterval(datasource, this.interval);
    }
  }

  // Grafana 6.2 (and older) is using this.interval for queries,
  // but Grafana 6.3+ is calculating interval again in queryRunner,
  // so we need to save-restore this.panel.interval.
  issueQueriesWithInterval(datasource: any, interval: any) {
    var origInterval = this.panel.interval;
    this.panel.interval = interval;
    var res = super.issueQueries(datasource);
    this.panel.interval = origInterval;
    return res;
  }

  onDataFramesReceived(data: DataFrame[]) {
    const dataList = this.processor.getSeriesList({
      dataList: data as DataFrame[],
      range: this.range,
    });
    this.onDataReceived(dataList);
  }

  onDataReceived(dataList: TimeSeries[]) {
    this.data = dataList;
    // Quick workaround for 7.0+. There is no call to
    // calculateInterval when enter Edit mode.
    if (!this.intervalMs) {
      this.calculateInterval();
    }

    let newBucketMatrix = this.convertDataToBuckets(
      dataList,
      this.range.from.valueOf(),
      this.range.to.valueOf(),
      this.intervalMs,
      true
    );

    this.bucketMatrix = newBucketMatrix;
    this.bucketMatrixPager.bucketMatrix = newBucketMatrix;
    if (newBucketMatrix.targets.length !== this.bucketMatrix.targets.length) {
      this.bucketMatrixPager.setCurrent(0);
    }

    this.noDatapoints = this.bucketMatrix.noDatapoints;

    if (this.annotationsPromise) {
      this.annotationsPromise.then(
        (result: { alertState: any; annotations: any }) => {
          this.loading = false;
          //this.alertState = result.alertState;
          if (result.annotations && result.annotations.length > 0) {
            this.annotations = result.annotations;
          } else {
            this.annotations = [];
          }
          this.render();
        },
        () => {
          this.loading = false;
          this.annotations = [];
          this.render();
        }
      );
    } else {
      this.loading = false;
      this.annotations = [];
      this.render();
    }
  }

  onInitEditMode() {
    this.addEditorTab('Options', optionsEditorCtrl, 2);
    this.addEditorTab('Tooltip', tooltipEditorCtrl, 3);
    this.addEditorTab('About', 'public/plugins/flant-statusmap-panel/partials/about_editor.html', 4);
    this.unitFormats = kbn.getUnitFormats();
  }

  // onRender will be called before StatusmapRenderer.onRender.
  // Decide if warning should be displayed over cards.
  onRender() {
    if (!this.range || !this.data) {
      this.noDatapoints = true;
      return;
    }

    this.multipleValues = false;
    if (!this.panel.useMax) {
      if (this.bucketMatrix) {
        this.multipleValues = this.bucketMatrix.multipleValues;
      }
    }

    this.noColorDefined = false;
    if (this.panel.color.mode === 'discrete') {
      if (this.panel.seriesFilterIndex === -1) {
        this.discreteExtraSeries.updateCardsValuesHasColorInfo();
      } else {
        this.discreteExtraSeries.updateCardsValuesHasColorInfoSingle();
      }
      if (this.bucketMatrix) {
        this.noColorDefined = this.bucketMatrix.noColorDefined;
      }
    }

    this.noDatapoints = false;
    if (this.bucketMatrix) {
      this.noDatapoints = this.bucketMatrix.noDatapoints;
    }
  }

  onCardColorChange(newColor) {
    this.panel.color.cardColor = newColor;
    this.render();
  }

  onDataError() {
    this.data = [];
    this.annotations = [];
    this.render();
  }

  postRefresh() {
    this.noColorDefined = false;
  }

  link(scope, elem, attrs, ctrl) {
    rendering(scope, elem, attrs, ctrl);
  }

  // Compatible with Grafana 7.0 overrides feature.
  retrieveTimeVar() {
    var time = this.timeSrv.timeRangeForUrl();
    return 'from=' + time.from + '&to=' + time.to;
  }

  // convertToBuckets groups values in data into buckets by target and timestamp.
  //
  // data is a result from datasource. It is an array of timeseries and tables:
  /* [
    // timeseries
    {
      target: "query alias",
      refId: "A",
      datapoints: [
        [0, 1582681239911],
        [3, 158....],
        ...
      ],
      tags?:{key: value,...}
    },
    // table
    {
      name: "table name",
      refId: "B",
      columns: [
        {text: "id"},
        {text: "info"},
        ...
      ],
      rows: [
        [1, "123"],
        [2, "44411"],
        ...
      ]
    },
...
  ]

  to and from — a time range of the panel.
  intervalMs — a calculated interval. It is used to split a time range.
  */
  convertDataToBuckets(
    data: any,
    from: number,
    to: number,
    intervalMs: number,
    mostRecentBucket: boolean
  ): BucketMatrix {
    let bucketMatrix = new BucketMatrix();
    bucketMatrix.rangeMs = to - from;
    bucketMatrix.intervalMs = intervalMs;

    if (!data || data.length === 0) {
      // Mimic heatmap and graph 'no data' labels.
      bucketMatrix.targets = ['1.0', '0.0', '-1.0'];
      bucketMatrix.buckets['1.0'] = [];
      bucketMatrix.buckets['0.0'] = [];
      bucketMatrix.buckets['-1.0'] = [];
      bucketMatrix.xBucketSize = 42;
      bucketMatrix.noDatapoints = true;
      return bucketMatrix;
    }

    let targetIndex: { [target: string]: number[] } = {};
    let targetPartials: { [target: string]: string[] } = {};

    // Group indicies of elements in data by target (y label).

    // lodash version:
    //_.map(data, (d, i) => {
    //  targetIndex[d.target] = _.concat(_.toArray(targetIndex[d.target]), i);
    //});

    data.map((queryResult: any, i: number) => {
      // Init yLabel as the query target
      let yLabel = queryResult.target;

      // Check if there is some labelTemplate configured
      if (this.panel.yLabel.usingSplitLabel && this.panel.yLabel.delimiter !== '') {
        let pLabels = queryResult.target.split(this.panel.yLabel.delimiter);

        // Load all possible values as scoped vars and load them into targetPartials
        // to be used on different components as Bucket and BucketMatrix props
        let scopedVars = [];
        scopedVars[`__y_label`] = { value: yLabel };
        for (let i in pLabels) {
          scopedVars[`__y_label_${i}`] = { value: pLabels[i] };
        }

        if (this.panel.yLabel.labelTemplate !== '') {
          yLabel = this.templateSrv.replace(this.panel.yLabel.labelTemplate, scopedVars);
        }

        targetPartials[yLabel] = pLabels;
      }

      //reset if it already exists
      if (!targetIndex.hasOwnProperty(yLabel)) {
        targetIndex[yLabel] = [];
      }
      targetIndex[yLabel].push(i);
    });

    let targetKeys = _.keys(targetIndex);

    //console.log ("targetIndex: ", targetIndex, "targetKeys: ", targetKeys);

    let targetTimestampRanges: { [target: string]: { [timestamp: number]: number[] } } = {};

    // Collect all timestamps for each target.
    // Make map timestamp => [from, to]. from == previous ts, to == ts from datapoint.
    targetKeys.map(target => {
      let targetTimestamps: any[] = [];

      for (let si = 0; si < targetIndex[target].length; si++) {
        let s = data[targetIndex[target][si]];
        _.map(s.datapoints, (datapoint, idx) => {
          targetTimestamps.push(datapoint[TIME_INDEX] - from);
        });
      }

      //console.log("timestamps['"+target+"'] = ", targetTimestamps);

      targetTimestamps = _.uniq(targetTimestamps);

      //console.log("uniq timestamps['"+target+"'] = ", targetTimestamps);

      targetTimestampRanges[target] = [];
      for (let i = targetTimestamps.length - 1; i >= 0; i--) {
        let tsTo = targetTimestamps[i];
        let tsFrom = 0;
        if (tsTo < 0) {
          tsFrom = tsTo - intervalMs;
        } else {
          if (i - 1 >= 0) {
            // Set from to previous timestamp + 1ms;
            tsFrom = targetTimestamps[i - 1] + 1;
            // tfTo - tfFrom should not be more than intervalMs
            let minFrom = tsTo - intervalMs;
            if (tsFrom < minFrom) {
              tsFrom = minFrom;
            }
          }
        }
        targetTimestampRanges[target][tsTo] = [tsFrom, tsTo];
      }
    });

    // console.log ("targetTimestampRanges: ", targetTimestampRanges);

    // Create empty buckets using intervalMs to calculate ranges.
    // If mostRecentBucket is set, create a bucket with a range "to":"to"
    // to store most recent values.
    targetKeys.map(target => {
      let targetEmptyBuckets: any[] = [];

      let lastTs = to - from;

      if (mostRecentBucket) {
        let topBucket = new Bucket();
        topBucket.yLabel = target;
        topBucket.relTo = lastTs;
        topBucket.relFrom = lastTs;
        topBucket.values = [];
        topBucket.mostRecent = true;
        if (targetTimestampRanges[target].hasOwnProperty(lastTs)) {
          topBucket.relFrom = targetTimestampRanges[target][lastTs][0];
          lastTs = topBucket.relFrom;
        }
        topBucket.to = topBucket.relTo + from;
        topBucket.from = topBucket.relFrom + from;
        targetEmptyBuckets.push(topBucket);
      }

      let idx = 0;
      let bucketFrom = 0;
      while (bucketFrom >= 0) {
        let b = new Bucket();
        b.yLabel = target;
        b.pLabels = targetPartials[target];
        b.relTo = lastTs - idx * intervalMs;
        b.relFrom = lastTs - (idx + 1) * intervalMs;
        b.to = b.relTo + from;
        b.from = b.relFrom + from;
        b.values = [];
        bucketFrom = b.relFrom;
        targetEmptyBuckets.push(b);
        idx++;
      }

      targetEmptyBuckets.map((bucket, i) => {
        bucket.xid = i;
      });

      bucketMatrix.buckets[target] = targetEmptyBuckets;
    });

    //console.log ("bucketMatrix: ", bucketMatrix);

    // Put values into buckets.
    bucketMatrix.minValue = Number.MAX_VALUE;
    bucketMatrix.maxValue = Number.MIN_SAFE_INTEGER;
    targetKeys.map(target => {
      targetIndex[target].map(dataIndex => {
        let s = data[dataIndex];
        s.datapoints.map((dp: any) => {
          for (let i = 0; i < bucketMatrix.buckets[target].length; i++) {
            if (bucketMatrix.buckets[target][i].belong(dp[TIME_INDEX])) {
              bucketMatrix.buckets[target][i].put(dp[VALUE_INDEX]);
            }
          }
        });
      });
      bucketMatrix.buckets[target].map(bucket => {
        bucket.minValue = _.min(bucket.values);
        bucket.maxValue = _.max(bucket.values);
        if (bucket.minValue < bucketMatrix.minValue) {
          bucketMatrix.minValue = bucket.minValue;
        }
        if (bucket.maxValue > bucketMatrix.maxValue) {
          bucketMatrix.maxValue = bucket.maxValue;
        }
        bucket.value = bucket.maxValue;
        if (bucket.values.length > 1) {
          bucketMatrix.multipleValues = true;
          bucket.multipleValues = true;

          bucket.value =
            this.panel.seriesFilterIndex !== -1 ? bucket.values[this.panel.seriesFilterIndex] : bucket.maxValue;
        }
      });
    });

    bucketMatrix.xBucketSize = Number.MIN_SAFE_INTEGER;
    targetKeys.map(target => {
      let bucketsLen: number = bucketMatrix.buckets[target].length;
      if (bucketsLen > bucketMatrix.xBucketSize) {
        bucketMatrix.xBucketSize = bucketsLen;
      }
    });

    //console.log ("bucketMatrix with values: ", bucketMatrix);

    bucketMatrix.targets = targetKeys;
    bucketMatrix.pLabels = targetPartials;
    return bucketMatrix;
  }
}

export { StatusHeatmapCtrl, StatusHeatmapCtrl as PanelCtrl };
